cpp-ch/local-engine/Common/GlutenSignalHandler.cpp

```diff
@@ -0,0 +1,422 @@
+            String calculated_binary_hash = getHashOfLoadedBinaryHex();
+            if (calculated_binary_hash == stored_binary_hash)
+            {
+                LOG_FATAL(log, "Integrity check of the executable successfully passed (checksum: {})", calculated_binary_hash);
+            }
+            else
+            {
+                LOG_FATAL(
+                    log,
+                    "Calculated checksum of the executable ({0}) does not correspond"
+                    " to the reference checksum stored in the executable ({1})."
+                    " This may indicate one of the following:"
+                    " - the executable was changed just after startup;"
+                    " - the executable was corrupted on disk due to faulty hardware;"
+                    " - the loaded executable was corrupted in memory due to faulty hardware;"
+                    " - the file was intentionally modified;"
+                    " - a logical error in the code.",
+                    calculated_binary_hash,
+                    stored_binary_hash);
+            }
+        }
+#endif
+        /// FIXME: Write crash to system.crash_log table if available.
+        if (collectGlutenCrashLog)
+            collectGlutenCrashLog(sig, thread_num, query_id, stack_trace);
+
+        ///TODO: Send crash report to developers (if configured)
+        if (sig != SanitizerTrap)
+        {
+            /// TODO: SentryWriter::onFault(sig, error_message, stack_trace);
+
+            /// TODO:  Advice the user to send it manually.
+        }
+        /// ClickHouse Keeper does not link to some part of Settings.
+#ifndef CLICKHOUSE_PROGRAM_STANDALONE_BUILD
+        /// List changed settings.
+        if (!query_id.empty())
+        {
+            ContextPtr query_context = thread_ptr->getQueryContext();
+            if (query_context)
+            {
+                String changed_settings = query_context->getSettingsRef().toString();
+
+                if (changed_settings.empty())
+                    LOG_FATAL(log, "No settings were changed");
+                else
+                    LOG_FATAL(log, "Changed settings: {}", changed_settings);
+            }
+        }
+#endif
+        /// When everything is done, we will try to send these error messages to client.
+        if (thread_ptr)
+            thread_ptr->onFatalError();
+
+        fatal_error_printed.test_and_set();
+    }
+};
+
+namespace local_engine
+{
+SignalHandler::SignalHandler() = default;
+SignalHandler::~SignalHandler() = default;
+
+struct SignalHandler::Impl
+{
+    /// A thread that acts on HUP and USR1 signal (close logs).
+    Poco::Thread signal_listener_thread;
+    std::unique_ptr<Poco::Runnable> signal_listener;
+    std::vector<int> handled_signals;
+
+    /// initialize termination process and signal handlers
+    void initializeTerminationAndSignalProcessing()
+    {
+        addSignalHandler({SIGABRT, SIGSEGV, SIGBUS, SIGTSTP}, signalHandler, &handled_signals);
```

Comment:
According this, https://man7.org/linux/man-pages/man7/signal.7.html, some signals causing core should also be handled.